ballerina/schemaClient.bal

```diff
@@ -0,0 +1,300 @@
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/sql;
+
+# Represents an SQL metadata client.
+isolated client class SchemaClient {
+    private final Client dbClient;
+    private final string database;
+
+    # Initializes the SchemaClient object
+    #
+    # + host - The name of where the mysql server is hosted (ex: localhost)
+    # + user - The username to access the database
+    # + password - The password to access the database
+    # + database - The name of the database to be accessed
+    # + return - A `sql:Error` or nil
+    public function init(string host, string user, string password, string database) returns sql:Error? {
+        self.database = database;
+        self.dbClient = check new (host, user, password);
+    }
+
+    # Retrieves all tables in the database.
+    #
+    # + return - A string array containing the names of the tables or an `sql:Error`
+    isolated remote function listTables() returns string[]|sql:Error {
+        string[] tables = [];
+        stream<record {|string table_name;|}, sql:Error?> results = self.dbClient->query(
+            `SELECT TABLE_NAME FROM INFORMATION_SCHEMA.TABLES 
+             WHERE TABLE_SCHEMA = ${self.database};`
+        );
+
+        do {
+            check from record {|string table_name;|} result in results
+                do {
+                    tables.push(result.table_name.toString());
+                };
+        } on fail error e {
+            return error(string `Error while listing the tables in  the ${self.database} database.`, cause = e);
+        }
+
+        do {
+            check results.close();
+        } on fail error e {
+            return error("Error while closing the result stream.", cause = e);
+        }
+
+        return tables;
+    }
+
+    # Retrieves information relevant to the provided table in the database.
+    #
+    # + tableName - The name of the table
+    # + include - Options on whether column and constraint related information should be fetched.
+    #             If `NO_COLUMNS` is provided, then no information related to columns will be retrieved.
+    #             If `COLUMNS_ONLY` is provided, then columnar information will be retrieved, but not constraint
+    #             related information.
+    #             If `COLUMNS_WITH_CONSTRAINTS` is provided, then columar information along with constraint related
+    #             information will be retrieved
+    # + return - An 'sql:TableDefinition' with the relevant table information or an `sql:Error`
+    isolated remote function getTableInfo(string tableName, sql:ColumnRetrievalOptions include = sql:COLUMNS_ONLY) returns sql:TableDefinition|sql:Error {
```

Comment:
This function seems a bit too long, and too many levels of indentation.

Would be better to break it down to several functions.
e.g. 
- `getColumns()`
- 'getConstraints()`